plugins/snapd/fu-snapd-observer.c

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright 2024 Maciej Borzecki <maciej.borzecki@canonical.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+#include "fu-snapd-observer.h"
+
+#include <curl/curl.h>
+#include <curl/easy.h>
+
+#include "fwupd-error.h"
+
+#include "fu-snapd-error.h"
+#include "fu-snapd-snap.h"
+
+struct _FuSnapdObserver {
+	GObject parent_instance;
+	CURL *curl_template;
+	struct curl_slist *req_hdrs;
+};
+
+G_DEFINE_TYPE(FuSnapdObserver, fu_snapd_observer, G_TYPE_OBJECT)
+
+FuSnapdObserver *
+fu_snapd_observer_new(void)
+{
+	return g_object_new(FU_TYPE_SNAPD_OBSERVER, NULL);
+}
+
+static void
+fu_snapd_observer_init(FuSnapdObserver *self)
+{
+	self->curl_template = curl_easy_init();
+	/* TODO choose different socket when in snap vs non-snap */
+	if (fu_snapd_is_in_snap()) {
+		curl_easy_setopt(self->curl_template,
+				 CURLOPT_UNIX_SOCKET_PATH,
+				 "/run/snapd-snap.socket");
+	} else {
+		curl_easy_setopt(self->curl_template,
+				 CURLOPT_UNIX_SOCKET_PATH,
+				 "/run/snapd.socket");
+	}
```

Comment:
this can't happen can it?  I thought you check in the `startup()` CB if you're running in the snap.